syntax = "proto3";

option java_multiple_files = true;
option java_package = "io.grpc.examples.raft_service";
option java_outer_classname = "RaftServiceProto";
option objc_class_prefix = "RSP";

package raft_service;

service RaftService {
  rpc AppendEntries(AppendEntriesRequest) returns (AppendEntriesReply) {}
  rpc RequestVote(RequestVoteRequest) returns (RequestVoteReply) {}
  rpc ClientWrite(ClientWriteRequest) returns (ClientWriteReply) {}
}

message ClientWriteRequest {
  string key = 1;
  string value = 2;
}

message AppendEntriesRequest {
  int32 term = 1;
  string leaderId = 2;
  int32 prevLogIndex = 3;
  int32 prevLogTerm = 4;
  repeated string entries = 5;
  int32 leaderCommit = 6;
}

message AppendEntriesReply {
  int32 term = 1;
  bool success = 2;
}

message ClientWriteReply {
  bool success = 1;
  string redirect = 2;
}

message RequestVoteRequest {
  int32 term = 1;
  string candidateId = 2;
  int32 lastLogIndex = 3;
  int32 lastLogTerm = 4;
}

message RequestVoteReply {
  int32 term = 1;
  bool voteGranted = 2;
}